homeassistant/components/broadlink/updater.py

```diff
@@ -0,0 +1,127 @@
+"""Support for fetching data from Broadlink devices."""
+from abc import ABC, abstractmethod
+from datetime import timedelta
+import logging
+
+from broadlink.exceptions import (
+    AuthorizationError,
+    BroadlinkException,
+    CommandNotSupportedError,
+    StorageError,
+)
+
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.util import dt
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def get_update_manager(device):
+    """Return an update manager for a given Broadlink device."""
+    update_managers = {
+        "A1": BroadlinkA1UpdateManager,
+        "MP1": BroadlinkMP1UpdateManager,
+        "RM2": BroadlinkRMUpdateManager,
+        "RM4": BroadlinkRMUpdateManager,
+        "SP1": BroadlinkSP1UpdateManager,
+        "SP2": BroadlinkSP2UpdateManager,
+    }
+    return update_managers[device.api.type](device)
+
+
+class BroadlinkUpdateManager(ABC):
+    """Representation of a Broadlink update manager.
+
+    Implement this class to manage fetching data from the device and to
+    monitor device availability.
+    """
+
+    def __init__(self, device):
+        """Initialize the update manager."""
+        self.device = device
+        self.coordinator = DataUpdateCoordinator(
+            device.hass,
+            _LOGGER,
+            name="device",
+            update_method=self.async_update,
+            update_interval=timedelta(minutes=1),
+        )
+        self.available = None
+        self.last_update = None
+
+    async def async_update(self):
+        """Fetch data from the device and update availability."""
+        try:
+            data = await self.async_fetch_data()
+
+        except BroadlinkException as err:
+            if self.available and (
+                dt.utcnow() - self.last_update > timedelta(minutes=3)
```

Comment:
Why should we not mark the entity as unavailable immediately? It failed an update.

Reply:
A communication failure does not always mean that the device is offline. Sometimes it is possible to control it right away. But not if it is marked as unavailable.

In the previous implementation we were not marking the device as unavailable when it failed an update. After I added this feature I realized that [it can bother some users](https://github.com/home-assistant/core/issues/37361#issue-650071204). Marking the device as unavailable is aggressive, as this will prevent the user from controlling it until the next successful update.

The idea here is to increase the fault tolerance. When something goes wrong, we will give the device at least 3 chances to recover spread over 3 minutes. If the device does not respond within 3 minutes, we will consider it unavailable. If it responds, we will accept it as a random failure and keep it available.

Reply:
Why do we think the device is able to respond to a command even though it failed an update?

Reply:
I intend to bring these hardcoded timedeltas to the options flow soon. So the user can choose the interval between updates and the timeout before marking the device as unavailable.

Reply:
I'd avoid making this logic user configurable. We should aim at making sane choices so the user don't have to worry about this.

Reply:
> Why do we think the device is able to respond to a command even though it failed an update?

Sometimes the update fails, even though the device is online. The cat may be passing in front of the router at the same time, or there may be a problem with routing that specific packet. This is common, especially when it comes to UDP.

It is annoying not to be able to send a code or check the temperature just because the last update failed. I don't want to be fooled with a temperature from hours ago, but in the last two minutes it probably hasn't changed much, so I wish I could see it on the frontend.

Reply:
I can decrease the tolerance to two minutes. What do you think?

Reply:
You know the device best. You should decide.

Reply:
Sorry about interjecting, my perspective might be helpful (I was one of the users that "it bothered" 😸).

For my use case, I was sending a command to the device about 6 times an hour and polling the sensors 60 times an hour. Every day or so, I would need to restart HA because of a single communication error. Maybe 1 in 100 commands would fail (my AP logged a migration from one AP to another). From a network administration perspective, a rate of 1% package loss is completely reasonable, even up to 3% wouldn't be cause for concern. I know the device was still available since it continued to make DHCP requests and respond to sensor queries from HA every minute.

Reply:
> I'd avoid making this logic user configurable. We should aim at making sane choices so the user don't have to worry about this.

Ok.

Reply:
@Silvenga Thanks for the feedback. That issue will disappear with this update. I've been testing this new version since the weekend. It is running smooth and without authorization failures. You are definitely gonna like it.